# MobiDB-lite: long disorder consensus predictor

MobiDB-lite is a Python program designed to generate a consensus prediction of intrinsically disordered regions in proteins.
The consensus is generated by measuring predictors agreement where at least 6/9 of predictors must agree to assign disorder state to a
residue. A mathematical morphology (MM) dilation/erosion processing is applied to smooth the prediction and short regions are filtered out.

The code is integrated into the [InterProScan](https://www.ebi.ac.uk/interpro/download/InterProScan/) software. 
The MobiDB-lite output is available in the 
[MobiDB](https://mobidb.org), 
[InterPro](https://www.ebi.ac.uk/interpro/), 
[PDBe](https://www.ebi.ac.uk/pdbe/), 
[PDBe-KB](https://www.ebi.ac.uk/pdbe/pdbe-kb/) and 
[UniProtKB](https://www.uniprot.org/) 
databases.

This version of MobiDB has been developed in collaboration with Matthias Blum who optimized its execution time by 10 times. 
Matthias' version is available [here](https://github.com/matthiasblum/idrpred.git) (and called IDRPred).

MobiDB-lite is also available as Docker container [MobiDB-lite_docker](https://github.com/BioComputingUP/MobiDB-lite_docker)

## Citation

[MobiDB-lite 3.0: Fast consensus annotation of intrinsic disorder flavors in proteins](https://dx.doi.org/10.1093/bioinformatics/btaa1045)  
*Necci M, Piovesan D, Clementel D, DosztÃ¡nyi Z, Tosatto SCE*  
**Bioinformatics (2020)** - DOI: [10.1093/bioinformatics/btaa1045](https://dx.doi.org/10.1093/bioinformatics/btaa1045)


## Installation
You can use the tool simply cloning this repo and running `__main__.py` Python script. 
which implements a command line interface.
To avoid import errors you need to add the source root folder 
to the `PYTHONPATH` environment variable. For example:

    export PYTHONPATH=/path/to/MobiDB-lite/src:$PYTHONPATH

Alternatively, you can install the package with Pip. No need to export any variable
in this case.

From GitHub:

    git clone https://github.com/BioComputingUP/MobiDB-lite.git
    pip install .


In the case the calculation of ensemble properties (compactness), the nu parameter, the following optional
dependencies are required:

* scikit-learn >= 1.4.2
* numba >= 0.59.1


## Usage

The program can be executing the command line interface or as a library.
Both the command line and the library accept the following positional arguments:

* **input file** - Protein sequences to be analyzed in FASTA format
* **output file** - The output file, different formats can be generated


### Command line

When executed from the command line the script logs information about the calculation in the console (standard error) and
will create an output file containing the prediction results. 
A different temporary folder can be specified using the `--tempdir` option. 

The following options are available:
When installed with pip the script is available as `mobidb_lite` command.

```bashcon
mobidb_lite input_fasta_file output_file 
```
If you simply cloned the repository:
```bashcon
python3 /path/to/MobiDB-lite/src/mobidb_lite/__main__.py input_fasta_file output_file 
```